internal/rows/arrowbased/arrowRows.go

```diff
@@ -0,0 +1,1158 @@
+	"github.com/databricks/databricks-sql-go/internal/config"
+	dbsqlerr "github.com/databricks/databricks-sql-go/internal/err"
+	"github.com/databricks/databricks-sql-go/internal/rows/rowscanner"
+	dbsqllog "github.com/databricks/databricks-sql-go/logger"
+	"github.com/pkg/errors"
+)
+
+var errArrowRowsUnsupportedNativeType = "databricks: arrow native values not yet supported for %s"
+var errArrowRowsUnsupportedWithHiveSchema = "databricks: arrow native values for %s require arrow schema"
+var errArrowRowsInvalidBatchIndex = "databricks: invalid arrow batch index index = %d, n = %d"
+var errArrowRowsNoArrowBatches = "databricks: result set contains 0 arrow batches"
+var errArrowRowsUnableToReadBatch = "databricks: unable to read arrow batch"
+var errArrowRowsNilArrowSchema = "databricks: nil arrow.Schema"
+var errArrowRowsUnableToWriteArrowSchema = "databricks: unable to write arrow schema"
+var errArrowRowsInvalidRowIndex = "databricks: row index %d is not contained in any arrow batch"
+var errArrowRowsInvalidDecimalType = "databricks: decimal type with no scale/precision"
+var errArrowRowsUnableToCreateDecimalType = "databricks: unable to create decimal type scale: %d, precision: %d"
+var errArrowRowsUnknownDBType = "unknown data type when converting to arrow type"
+var errArrowRowsUnhandledArrowType = "databricks: arrow row scanner unhandled type %s"
+var errArrowRowsDateTimeParse = "databrics: arrow row scanner failed to parse date/time"
+
+type recordReader interface {
+	NewRecordFromBytes(arrowSchemaBytes []byte, sparkArrowBatch sparkArrowBatch) (arrow.Record, error)
+}
+
+type valueContainerMaker interface {
+	makeColumnValuesContainers(ars *arrowRowScanner) error
+}
+
+type sparkArrowBatch struct {
+	rowCount, startRow, endRow int64
+	arrowRecordBytes           []byte
+	index                      int
+}
+
+func (sab *sparkArrowBatch) contains(rowIndex int64) bool {
+	return sab != nil && sab.startRow <= rowIndex && sab.endRow >= rowIndex
+}
+
+type timeStampFn func(arrow.Timestamp) time.Time
+
+type colInfo struct {
+	name      string
+	arrowType arrow.DataType
+	dbType    cli_service.TTypeId
+}
+
+// arrowRowScanner handles extracting values from arrow records
+type arrowRowScanner struct {
+	recordReader
+	valueContainerMaker
+
+	// configuration of different arrow options for retrieving results
+	config.ArrowConfig
+
+	// arrow batches returned by the thrift server
+	arrowBatches []sparkArrowBatch
+
+	// arrow schema corresponding to the TTableSchema
+	arrowSchema *arrow.Schema
+
+	// serialized form of arrow format schema
+	arrowSchemaBytes []byte
+
+	// database types for the columns
+	colInfo []colInfo
+
+	// number of rows in the current TRowSet
+	nRows int64
+
+	// a TRowSet contains multiple arrow batches
+	currentBatch *sparkArrowBatch
+
+	// Values for each column
+	columnValues []columnValues
+
+	// function to convert arrow timestamp when using native arrow format
+	toTimestampFn timeStampFn
+
+	*dbsqllog.DBSQLLogger
```

Comment:
why does this have a logger?

Reply:
Holding on to a logger passed to the NewArrowRows function, rather than just using the global from dbsqllog, so that the logger can have additional context of correlation ID and operation ID.
I will add some comments to make this clear.